In [40]:
import pandas as pd
import requests
import json
import tweepy
from tweepy import OAuthHandler
from timeit import default_timer as timer
import matplotlib.pyplot as plt
import os
import sys
%matplotlib inline

## Gather

#### twitter-archive-enhanced.csv

In [10]:
tweets_df=pd.read_csv('twitter-archive-enhanced.csv')

#### image-predictions.tsv

In [2]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
file_name ='image-predictions.tsv'
response = requests.get(url)

In [5]:
#if not os.path.isfile(file_name):
with open(file_name, 'wb') as f:
    f.write(response.content)

In [7]:
prdictions_df=pd.read_csv('image-predictions.tsv',sep='	')

#### Twitter 

In [22]:

consumer_key = 'YQGytf2YFV9xtAE8AX2gOUXp5'
consumer_secret = 'T1ikEvFYCD07khhU9iVJAbRUrp9i1FkvgIk6Cwqef7uqVlTAMT'
access_token = '434798336-5a7g2pN1gMJEdosVbjeUHl91lkxGQJctHGAAUpKR'
access_secret = 'WgVHwZpHbiG3LQTetswcjkUyAKIQnnZfxXZXSz6IxnDuk'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

df_1 = tweets_df.copy()
# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = df_1.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open(os.path.join(sys.path[0],'tweet_json.txt'), 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

1: 892420643555336193
Success
2: 892177421306343426
Success
3: 891815181378084864
Success
4: 891689557279858688
Success
5: 891327558926688256
Success
6: 891087950875897856
Success
7: 890971913173991426
Success
8: 890729181411237888
Success
9: 890609185150312448
Success
10: 890240255349198849
Success
11: 890006608113172480
Success
12: 889880896479866881
Success
13: 889665388333682689
Success
14: 889638837579907072
Success
15: 889531135344209921
Success
16: 889278841981685760
Success
17: 888917238123831296
Success
18: 888804989199671297
Success
19: 888554962724278272
Success
20: 888202515573088257
Fail
21: 888078434458587136
Success
22: 887705289381826560
Success
23: 887517139158093824
Success
24: 887473957103951883
Success
25: 887343217045368832
Success
26: 887101392804085760
Success
27: 886983233522544640
Success
28: 886736880519319552
Success
29: 886680336477933568
Success
30: 886366144734445568
Success
31: 886267009285017600
Success
32: 886258384151887873
Success
33: 8860541600590725

Fail
262: 842846295480000512
Success
263: 842765311967449089
Success
264: 842535590457499648
Success
265: 842163532590374912
Success
266: 842115215311396866
Success
267: 841833993020538882
Success
268: 841680585030541313
Success
269: 841439858740625411
Success
270: 841320156043304961
Success
271: 841314665196081154
Success
272: 841077006473256960
Success
273: 840761248237133825
Success
274: 840728873075638272
Success
275: 840698636975636481
Success
276: 840696689258311684
Success
277: 840632337062862849
Success
278: 840370681858686976
Success
279: 840268004936019968
Success
280: 839990271299457024
Success
281: 839549326359670784
Success
282: 839290600511926273
Success
283: 839239871831150596
Success
284: 838952994649550848
Success
285: 838921590096166913
Success
286: 838916489579200512
Success
287: 838831947270979586
Success
288: 838561493054533637
Success
289: 838476387338051585
Success
290: 838201503651401729
Success
291: 838150277551247360
Success
292: 838085839343206401
Success
293

Success
519: 810657578271330305
Success
520: 810284430598270976
Success
521: 810254108431155201
Success
522: 809920764300447744
Success
523: 809808892968534016
Success
524: 809448704142938112
Success
525: 809220051211603969
Success
526: 809084759137812480
Success
527: 808838249661788160
Success
528: 808733504066486276
Success
529: 808501579447930884
Success
530: 808344865868283904
Success
531: 808134635716833280
Success
532: 808106460588765185
Success
533: 808001312164028416
Success
534: 807621403335917568
Success
535: 807106840509214720
Success
536: 807059379405148160
Success
537: 807010152071229440
Success
538: 806629075125202948
Success
539: 806620845233815552
Success
540: 806576416489959424
Success
541: 806542213899489280
Success
542: 806242860592926720
Success
543: 806219024703037440
Success
544: 805958939288408065
Success
545: 805932879469572096
Success
546: 805826884734976000
Success
547: 805823200554876929
Success
548: 805520635690676224
Success
549: 805487436403003392
Success


Success
776: 776201521193218049
Success
777: 776113305656188928
Success
778: 776088319444877312
Success
779: 775898661951791106
Success
780: 775842724423557120
Success
781: 775733305207554048
Success
782: 775729183532220416
Success
783: 775364825476165632
Success
784: 775350846108426240
Success
785: 775096608509886464
Fail
786: 775085132600442880
Success
787: 774757898236878852
Success
788: 774639387460112384
Success
789: 774314403806253056
Success
790: 773985732834758656
Success
791: 773922284943896577
Success
792: 773704687002451968
Success
793: 773670353721753600
Success
794: 773547596996571136
Success
795: 773336787167145985
Success
796: 773308824254029826
Success
797: 773247561583001600
Success
798: 773191612633579521
Success
799: 772877495989305348
Success
800: 772826264096874500
Success
801: 772615324260794368
Success
802: 772581559778025472
Success
803: 772193107915964416
Success
804: 772152991789019136
Success
805: 772117678702071809
Success
806: 772114945936949249
Success
807

Fail
1045: 743609206067040256
Fail
1046: 743595368194129920
Fail
1047: 743545585370791937
Fail
1048: 743510151680958465
Fail
1049: 743253157753532416
Fail
1050: 743222593470234624
Fail
1051: 743210557239623680
Fail
1052: 742534281772302336
Fail
1053: 742528092657332225
Fail
1054: 742465774154047488
Fail
1055: 742423170473463808
Fail
1056: 742385895052087300
Fail
1057: 742161199639494656
Fail
1058: 742150209887731712
Fail
1059: 741793263812808706
Fail
1060: 741743634094141440
Fail
1061: 741438259667034112
Fail
1062: 741303864243200000
Fail
1063: 741099773336379392
Fail
1064: 741067306818797568
Fail
1065: 740995100998766593
Fail
1066: 740711788199743490
Fail
1067: 740699697422163968
Fail
1068: 740676976021798912
Fail
1069: 740373189193256964
Fail
1070: 740365076218183684
Fail
1071: 740359016048689152
Fail
1072: 740214038584557568
Fail
1073: 739979191639244800
Fail
1074: 739932936087216128
Fail
1075: 739844404073074688
Fail
1076: 739623569819336705
Fail
1077: 739606147276148736
Fail
1078:

Fail
1319: 706538006853918722
Fail
1320: 706516534877929472
Fail
1321: 706346369204748288
Fail
1322: 706310011488698368
Fail
1323: 706291001778950144
Fail
1324: 706265994973601792
Fail
1325: 706169069255446529
Fail
1326: 706166467411222528
Fail
1327: 706153300320784384
Fail
1328: 705975130514706432
Fail
1329: 705970349788291072
Fail
1330: 705898680587526145
Fail
1331: 705786532653883392
Fail
1332: 705591895322394625
Fail
1333: 705475953783398401
Fail
1334: 705442520700944385
Fail
1335: 705428427625635840
Fail
1336: 705239209544720384
Fail
1337: 705223444686888960
Fail
1338: 705102439679201280
Fail
1339: 705066031337840642
Fail
1340: 704871453724954624
Fail
1341: 704859558691414016
Fail
1342: 704847917308362754
Fail
1343: 704819833553219584
Fail
1344: 704761120771465216
Fail
1345: 704499785726889984
Fail
1346: 704491224099647488
Fail
1347: 704480331685040129
Fail
1348: 704364645503647744
Fail
1349: 704347321748819968
Fail
1350: 704134088924532736
Fail
1351: 704113298707505153
Fail
1352:

Fail
1593: 686394059078897668
Fail
1594: 686386521809772549
Fail
1595: 686377065986265092
Fail
1596: 686358356425093120
Fail
1597: 686286779679375361
Fail
1598: 686050296934563840
Fail
1599: 686035780142297088
Fail
1600: 686034024800862208
Fail
1601: 686007916130873345
Fail
1602: 686003207160610816
Fail
1603: 685973236358713344
Fail
1604: 685943807276412928
Fail
1605: 685906723014619143
Fail
1606: 685681090388975616
Fail
1607: 685667379192414208
Fail
1608: 685663452032069632
Fail
1609: 685641971164143616
Fail
1610: 685547936038666240
Fail
1611: 685532292383666176
Fail
1612: 685325112850124800
Fail
1613: 685321586178670592
Fail
1614: 685315239903100929
Fail
1615: 685307451701334016
Fail
1616: 685268753634967552
Fail
1617: 685198997565345792
Fail
1618: 685169283572338688
Fail
1619: 684969860808454144
Fail
1620: 684959798585110529
Fail
1621: 684940049151070208
Fail
1622: 684926975086034944
Fail
1623: 684914660081053696
Fail
1624: 684902183876321280
Fail
1625: 684880619965411328
Fail
1626:

Success
1851: 675740360753160193
Success
1852: 675710890956750848
Success
1853: 675707330206547968
Success
1854: 675706639471788032
Success
1855: 675534494439489536
Success
1856: 675531475945709568
Success
1857: 675522403582218240
Success
1858: 675517828909424640
Success
1859: 675501075957489664
Success
1860: 675497103322386432
Success
1861: 675489971617296384
Success
1862: 675483430902214656
Success
1863: 675432746517426176
Success
1864: 675372240448454658
Success
1865: 675362609739206656
Success
1866: 675354435921575936
Success
1867: 675349384339542016
Success
1868: 675334060156301312
Success
1869: 675166823650848770
Success
1870: 675153376133427200
Success
1871: 675149409102012420
Success
1872: 675147105808306176
Success
1873: 675146535592706048
Success
1874: 675145476954566656
Success
1875: 675135153782571009
Success
1876: 675113801096802304
Success
1877: 675111688094527488
Success
1878: 675109292475830276
Success
1879: 675047298674663426
Success
1880: 675015141583413248
Success
18

Success
2100: 670717338665226240
Fail
2101: 670704688707301377
Success
2102: 670691627984359425
Success
2103: 670679630144274432
Success
2104: 670676092097810432
Success
2105: 670668383499735048
Success
2106: 670474236058800128
Success
2107: 670468609693655041
Fail
2108: 670465786746662913
Success
2109: 670452855871037440
Success
2110: 670449342516494336
Success
2111: 670444955656130560
Success
2112: 670442337873600512
Success
2113: 670435821946826752
Success
2114: 670434127938719744
Success
2115: 670433248821026816
Success
2116: 670428280563085312
Success
2117: 670427002554466305
Success
2118: 670421925039075328
Success
2119: 670420569653809152
Success
2120: 670417414769758208
Success
2121: 670411370698022913
Success
2122: 670408998013820928
Success
2123: 670403879788544000
Success
2124: 670385711116361728
Success
2125: 670374371102445568
Success
2126: 670361874861563904
Success
2127: 670338931251150849
Success
2128: 670319130621435904
Success
2129: 670303360680108032
Success
2130: 67

Success
2349: 666055525042405380
Success
2350: 666051853826850816
Success
2351: 666050758794694657
Success
2352: 666049248165822465
Success
2353: 666044226329800704
Success
2354: 666033412701032449
Success
2355: 666029285002620928
Success
2356: 666020888022790149
Success
1363.7026446000018
{888202515573088257: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 875747767867523072: TweepError('Failed to send request: HTTPSConnectionPool(host=\'api.twitter.com\', port=443): Max retries exceeded with url: /1.1/statuses/show.json?id=875747767867523072&tweet_mode=extended (Caused by SSLError(SSLError("bad handshake: SysCallError(10054, \'WSAECONNRESET\')")))'), 873697596434513921: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 872668790621863937: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 872261713294495745: TweepError([{'code': 144, 'message': 'No status found with that ID.'}]), 869988702071779329: TweepError([{'c

In [23]:
tweet_id=[]
favorite_count=[]
retweet_count=[]

with open('tweet_json.txt', 'r') as r:
        for line in r:
            parse = json.loads(line)
            tweet_id.append(parse['id_str'])
            favorite_count.append(parse['favorite_count'])
            retweet_count.append(parse['retweet_count'])
            #print(parse['id'],'-',parse['favorite_count'],'-',parse['retweet_count'])

In [24]:
api_df = pd.DataFrame(columns=['tweet_id', 'favorite_count','retweet_count'])
api_df['tweet_id']=tweet_id
api_df['favorite_count']=favorite_count
api_df['retweet_count']=retweet_count

## Assess

#### Tweets

In [27]:
tweets_df.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [39]:
tweets_df.sample(30)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
868   761750502866649088                    NaN                  NaN   
2114  670433248821026816                    NaN                  NaN   
2137  670046952931721218                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
491   813800681631023104                    NaN                  NaN   
2042  671538301157904385                    NaN                  NaN   
1044  743609206067040256                    NaN                  NaN   
573   801167903437357056                    NaN                  NaN   
2167  669359674819481600                    NaN                  NaN   
1484  693231807727280129                    NaN                  NaN   
1417  698549713696649216                    NaN                  NaN   
138   865718153858494464                    NaN                  NaN   
1561  688385280030670848                    NaN                  NaN   
1097  736225175608430592                    NaN                  NaN   
108   871515927908634625                    NaN                  NaN   
304   836380477523124226                    NaN                  NaN   
1108  734559631394082816                    NaN                  NaN   
761   778286810187399168                    NaN                  NaN   
418   822462944365645825                    NaN                  NaN   
320   834458053273591808                    NaN                  NaN   
1803  676948236477857792                    NaN                  NaN   
871   761599872357261312                    NaN                  NaN   
875   761227390836215808                    NaN                  NaN   
2175  669214165781868544                    NaN                  NaN   
1713  680473011644985345                    NaN                  NaN   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1220  714606013974974464                    NaN                  NaN   
181   857029823797047296                    NaN                  NaN   
548   805487436403003392                    NaN                  NaN   

                      timestamp  \
868   2016-08-06 02:27:27 +0000   
2114  2015-11-28 02:45:17 +0000   
2137  2015-11-27 01:10:17 +0000   
17    2017-07-22 16:56:37 +0000   
2350  2015-11-16 00:30:50 +0000   
491   2016-12-27 17:36:16 +0000   
2042  2015-12-01 03:56:22 +0000   
1044  2016-06-17 01:00:24 +0000   
573   2016-11-22 20:58:07 +0000   
2167  2015-11-25 03:39:17 +0000   
1484  2016-01-30 00:38:37 +0000   
1417  2016-02-13 16:50:04 +0000   
138   2017-05-19 23:57:46 +0000   
1561  2016-01-16 15:40:14 +0000   
1097  2016-05-27 15:58:54 +0000   
108   2017-06-04 23:56:03 +0000   
304   2017-02-28 01:00:19 +0000   
1108  2016-05-23 01:40:38 +0000   
761   2016-09-20 17:36:50 +0000   
418   2017-01-20 15:17:01 +0000   
320   2017-02-22 17:41:18 +0000   
1803  2015-12-16 02:13:31 +0000   
871   2016-08-05 16:28:54 +0000   
875   2016-08-04 15:48:47 +0000   
2175  2015-11-24 18:01:05 +0000   
1713  2015-12-25 19:39:43 +0000   
1295  2016-03-10 17:35:20 +0000   
1220  2016-03-29 00:12:05 +0000   
181   2017-04-26 00:33:27 +0000   
548   2016-12-04 19:02:24 +0000   

                                                 source  \
868   <a href="http://twitter.com/download/iphone" r...   
2114  <a href="http://twitter.com/download/iphone" r...   
2137  <a href="http://twitter.com/download/iphone" r...   
17    <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
491   <a href="http://twitter.com/download/iphone" r...   
2042  <a href="http://twitter.com/download/iphone" r...   
1044  <a href="http://twitter.com/download/iphone" r...   
573   <a href="http://twitter.com/download/iphone" r...   
2167  <a href="http://twitter.com/download/iphone" r...   
1484  <a href="http://twitter.com/download/iphon

In [30]:
tweets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [46]:
tweets_df['doggo'].unique()

array(['None', 'doggo'], dtype=object)

In [42]:
#plot = tweets_df.plot.pie(y='rating_numerator', figsize=(5, 5))
tweets_df['rating_numerator'].unique()

array([  13,   12,   14,    5,   17,   11,   10,  420,  666,    6,   15,
        182,  960,    0,   75,    7,   84,    9,   24,    8,    1,   27,
          3,    4,  165, 1776,  204,   50,   99,   80,   45,   60,   44,
        143,  121,   20,   26,    2,  144,   88], dtype=int64)

In [44]:
tweets_df['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [20]:
prdictions_df.head(1)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   

   img_num                      p1   p1_conf  p1_dog      p2   p2_conf  \
0        1  Welsh_springer_spaniel  0.465074    True  collie  0.156665   

   p2_dog                 p3   p3_conf  p3_dog  
0    True  Shetland_sheepdog  0.061428    True

In [53]:
prdictions_df.sample(60)

tweet_id                                            jpg_url  \
992   708109389455101952    https://pbs.twimg.com/media/CdO1u9vWAAApj2V.jpg   
1930  858843525470990336    https://pbs.twimg.com/media/C-s5oYZXkAAMHHq.jpg   
607   680055455951884288    https://pbs.twimg.com/media/CW-ZRC_WQAAyFrL.jpg   
1392  767122157629476866    https://pbs.twimg.com/media/CqVdiBJWIAEDZB4.jpg   
1823  834931633769889797    https://pbs.twimg.com/media/C5ZF4p-XEAEmApg.jpg   
1669  813066809284972545    https://pbs.twimg.com/media/C0iX8OOVEAEIpMC.jpg   
1801  831939777352105988    https://pbs.twimg.com/media/C4uk0EWWQAAaZm1.jpg   
1614  802239329049477120    https://pbs.twimg.com/media/CyIgaTEVEAA-9zS.jpg   
487   675501075957489664    https://pbs.twimg.com/media/CV_cnjHWUAADc-c.jpg   
461   674800520222154752    https://pbs.twimg.com/media/CV1ffl3XAAAiFyr.jpg   
432   674291837063053312    https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
675   683449695444799489    https://pbs.twimg.com/media/CXwZ3pbWsAAriTv.jpg   
1448  776113305656188928    https://pbs.twimg.com/media/CsVO7ljW8AAckRD.jpg   
122   668221241640230912    https://pbs.twimg.com/media/CUX_rAyWsAYZOQ5.jpg   
83    667495797102141441    https://pbs.twimg.com/media/CUNr4-7UwAAg2lq.jpg   
1973  869702957897576449    https://pbs.twimg.com/media/DBHOOfOXoAABKlU.jpg   
1870  844704788403113984    https://pbs.twimg.com/media/C7j-hkSW0AIxCZC.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
228   670385711116361728    https://pbs.twimg.com/media/CU2wPyWWUAAb1MJ.jpg   
128   668274247790391296    https://pbs.twimg.com/media/CUYv4d2WUAAziXs.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
449   674737130913071104    https://pbs.twimg.com/media/CV0l10AU8AAfg-a.jpg   
941   704054845121142784    https://pbs.twimg.com/media/CcVOJEcXEAM0FHL.jpg   
2017  880095782870896641    https://pbs.twimg.com/media/DDa6ckbXgAAM1vV.jpg   
580   678767140346941444    https://pbs.twimg.com/media/CWt3G6EVEAIGEPr.jpg   
1922  856543823941562368    https://pbs.twimg.com/media/C-MOEDCXYAEjp7o.jpg   
1057  714631576617938945    https://pbs.twimg.com/media/CerhoBWWAAA5eLL.jpg   
1047  713177543487135744    https://pbs.twimg.com/media/CeW3MWMWQAEOMbq.jpg   
903   700062718104104960    https://pbs.twimg.com/media/CbcfUxoUAAAlHGK.jpg   
1603  800141422401830912    https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
867   697575480820686848    https://pbs.twimg.com/media/Ca5JMvMUsAAGMll.jpg   
1339  758474966123810816    https://pbs.twimg.com/media/Coak48zWAAAhBxV.jpg   
1274  750041628174217216    https://pbs.twimg.com/media/CmfssOtXYAAKa_Z.jpg   
1507  785639753186217984    https://pbs.twimg.com/media/CucnLmeWAAALOSC.jpg   
1385  766008592277377025    https://pbs.twimg.com/media/CqFouXOXYAAYpzG.jpg   
442   674632714662858753    https://pbs.twimg.com/media/CVzG3yOVAAAqi9I.jpg   
1451  776477788987613185    https://pbs.twimg.com/media/CsaaaaxWgAEfzM7.jpg   
1095  720043174954147842    https://pbs.twimg.com/media/Cf4bcm8XEAAX4xV.jpg   
1148  731156023742988288    https://pbs.twimg.com/media/CiWWhVNUYAAab_r.jpg   
1376  763183847194451968    https://pbs.twimg.com/media/CpdfpzKWYAAWSUi.jpg   
72    667211855547486208    https://pbs.twimg.com/media/CUJppKJWoAA75NP.jpg   
1429  772877495989305348  https://pbs.twimg.com/ext_tw_video_thumb/77287...   
1482  781163403222056960    https://pbs.twimg.com/media/Ctc_-BTWEAAQpZh.jpg   
1945  862096992088072192    https://pbs.twimg.com/media/C_bIo7QXYAAGfPu.jpg   
891   699370870310113280    https://pbs.twimg.com/media/CbSqE0rVIAEOPE4.jpg   
309   671538301157904385    https://pbs.twimg.com/media/CVHIhi2WsAEgdKk.jpg   
1140  729823566028484608    https://pbs.twimg.com/media/CiDap8fWEAAC4iW.jpg   
816   692828166163931137    https://pbs.twimg.com/media/CZ1riVOWwAATfGf.jpg   
205   669923323644657664    https://pbs.twimg.com/media/CUwLtPeU8A

In [54]:
prdictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [28]:
api_df.head(1)

tweet_id  favorite_count  retweet_count
0  892420643555336193           35695           7549

### Quality

- There are 181 retweeted tweets  ✔ 
- There are 78 replied tweets  ✔ 
- tweet_id is of int type and won't be need for and statistics  ✔ 
- timestamp is of string type  ✔ 
- there are rating_numerator less than 10 while the main theme is that all ratings are R >=10
- None insted NaN in the following coulmns (doggo, floofer, pupper, puppo)
- some rows has False in the 3 prdictions which means its not a dog
- there is discrepancy in the counts of tweets and the counts image predictions which needs to be resolved

### Tidiness 

- Columns (doggo, floofer, pupper, puppo) are values not varible names
- After cleaning(removing) non-orginal tweets the 5 columns (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)  won't be needed.  ✔ 

## Cleaning

In [56]:
tweets_df_clean=tweets_df.copy()

#### Define

- Get rid of the 181 retweeted tweets
- Get rid of the 78 replied tweets

In [58]:
df_no_relply=tweets_df_clean[tweets_df_clean['in_reply_to_status_id'].isnull()]

In [59]:
df_no_relply_and_retweets=df_no_relply[df_no_relply['retweeted_status_id'].isnull()]

#### Test

In [60]:
df_no_relply_and_retweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2097 non-null   int64  
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2097 non-null   object 
 4   source                      2097 non-null   object 
 5   text                        2097 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2094 non-null   object 
 10  rating_numerator            2097 non-null   int64  
 11  rating_denominator          2097 non-null   int64  
 12  name                        2097 non-null   object 
 13  doggo                       2097 

#### Define

- After cleaning(removing) non-orginal tweets the 5 columns (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp) won't be needed.

#### Code

In [ ]:
df_no_relply_and_retweets.drop(['in_reply_to_status_id', 'in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],inplace=True, axis=1)

#### Test

In [67]:
list(df_no_relply_and_retweets)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

#### Define 
- timestamp from string DateTime
- convert Tweet ID from int to String

In [ ]:
df_no_relply_and_retweets['timestamp'] = pd.to_datetime(df_no_relply_and_retweets['timestamp'].str[:-6], format='%Y-%m-%d %H:%M:%S')

In [81]:
df_no_relply_and_retweets=df_no_relply_and_retweets.astype({'tweet_id': 'str'})

#### Test 

In [82]:
df_no_relply_and_retweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   tweet_id            2097 non-null   object        
 1   timestamp           2097 non-null   datetime64[ns]
 2   source              2097 non-null   object        
 3   text                2097 non-null   object        
 4   expanded_urls       2094 non-null   object        
 5   rating_numerator    2097 non-null   int64         
 6   rating_denominator  2097 non-null   int64         
 7   name                2097 non-null   object        
 8   doggo               2097 non-null   object        
 9   floofer             2097 non-null   object        
 10  pupper              2097 non-null   object        
 11  puppo               2097 non-null   object        
dtypes: datetime64[ns](1), int64(2), object(9)
memory usage: 213.0+ KB


#### Define

- Get ride of tweets with ratings less than 10 

#### Code

In [83]:
clean_tweet_df=df_no_relply_and_retweets.copy()

In [84]:
clean_tweet_df=clean_tweet_df[(clean_tweet_df.rating_numerator >= 10)]

In [85]:
clean_tweet_df['rating_numerator'].unique()

array([  13,   12,   14,   11,   10,   84,   24,   75,   27,  165, 1776,
        204,   50,   99,   80,   45,   60,   44,  121,   26,  144,   88,
        420], dtype=int64)

In [86]:
clean_tweet_df['rating_denominator'].unique()

array([ 10,  70,   7, 150, 170,  50,  90,  80,  40, 110, 120], dtype=int64)

In [87]:
clean_tweet_df['rating_denominator'].value_counts()

10     1665
50        3
80        2
170       1
150       1
120       1
110       1
90        1
70        1
40        1
7         1
Name: rating_denominator, dtype: int64

In [97]:
x=clean_tweet_df[(clean_tweet_df.rating_denominator != 10)]

In [98]:
x.to_csv('inspect.csv')